### DISCUSSION TITLE
RESTFull API

### DISCUSSION THREAD

**Discussion by lperdereau** [cloud-hypervisor/cloud-hypervisor#3284] [Q&A]
Hello,

When I look Rest API of cloud-hypervisor, I find usage of HTTP verb poor.
Could it be possible to make this type of API ?

#### Virtual Machine Manager (VMM) Actions

Action                              | HTTP Verbs  | Endpoint     | Request Body | Response Body              | Prerequisites
------------------------------------|-------------|--------------|--------------|----------------------------|---------------------
Check for the REST API availability | OPTIONS     | `/vmm`       | N/A          | `/schemas/VmmPingResponse` | N/A
Shut the VMM down                   | DELETE      | `/vmm`       | N/A          | N/A                        | The VMM is running

#### Virtual Machine (VM) Actions

Action                             | Endpoint     | Request Body         | Response Body             | Prerequisites
-----------------------------------|--------------|----------------------|---------------------------|---------------------------
Create the VM                      | POST         | `/vm`                | `/schemas/VmConfig`        | N/A                      | The VM is not created yet
Delete the VM                      | DELETE       | `/vm`                | N/A                       | N/A                      | N/A
Boot the VM                        | PUT          | `/vm/boot`           | N/A                       | N/A                      | The VM is created but not booted
Shut the VM down                   | PUT          | `/vm/shutdown`       | N/A                       | N/A                      | The VM is booted
Reboot the VM                      | PUT          | `/vm/reboot`         | N/A                       | N/A                      | The VM is booted
Trigger power button of the VM     | PUT          | `/vm/power-button`   | N/A                       | N/A                      | The VM is booted
Pause the VM                       | PUT          | `/vm/pause`          | N/A                       | N/A                      | The VM is booted
Resume the VM                      | PUT          | `/vm/resume`         | N/A                       | N/A                      | The VM is paused
Add/remove CPUs to/from the VM     | PUT          | `/vm/resize`         | `/schemas/VmResize`       | N/A                      | The VM is booted
Add/remove memory from a zone      | PUT          | `/vm/resize-zone`    | `/schemas/VmResizeZone`   | N/A                      | The VM is booted
Dump the VM information            | GET          | `/vm`                | N/A                       | `/schemas/VmInfo`        | The VM is created
Add VFIO PCI device to the VM      | POST         | `/vm/device`         | `/schemas/VmAddDevice`    | `/schemas/PciDeviceInfo` | The VM is booted
Add disk device to the VM          | POST         | `/vm/disk`           | `/schemas/DiskConfig`      | `/schemas/PciDeviceInfo` | The VM is booted
Add fs device to the VM            | POST         | `/vm/fs`             | `/schemas/FsConfig`        | `/schemas/PciDeviceInfo` | The VM is booted
Add pmem device to the VM          | POST         | `/vm/pmem`           | `/schemas/PmemConfig`      | `/schemas/PciDeviceInfo` | The VM is booted
Add network device to the VM       | POST         | `/vm/net`            | `/schemas/NetConfig`       | `/schemas/PciDeviceInfo` | The VM is booted
Add userspace PCI device to the VM | POST         | `/vm/user-device`    | `/schemas/VmAddUserDevice`| `/schemas/PciDeviceInfo` | The VM is booted
Add vsock device to the VM         | POST         | `/vm/vsock`          | `/schemas/VsockConfig`     | `/schemas/PciDeviceInfo` | The VM is booted
Remove device from the VM          | DELETE       | `/vm/device`         | `/schemas/VmRemoveDevice` | N/A                      | The VM is booted
Dump the VM counters               | GET          | `/vm/counters`       | N/A                       | `/schemas/VmCounters`    | The VM is booted

**Comment by rbradford** [comment]
Thank you for taking the time to analyse the API and propose changes. If you would like to take this work forward I propose the following steps:

Creating a PR that:
* Adds an `api-v2.md` file describing the new API
* Adds a `cloud-hypervisior-v2.yaml` OpenAPI YAML file for that API (Kata uses this for autogeneration)
* Introduce new code for the v2 API endpoint based on your detailed description above.
* Ports `ch-remote` to the new API

We can then work with the uses of the v1 API to move them over to the new API, then deprecate the v1 API and ultimately remove it 2 releases after its deprecation.
